pkg/backup/resource_backupper.go

```diff
@@ -0,0 +1,255 @@
+	labelSelector string,
+	dynamicFactory client.DynamicFactory,
+	discoveryHelper discovery.Helper,
+	backedUpItems map[itemKey]struct{},
+	actions map[schema.GroupResource]Action,
+	podCommandExecutor podCommandExecutor,
+	tarWriter tarWriter,
+	resourceHooks []resourceHook,
+) resourceBackupper {
+	return &defaultResourceBackupper{
+		log:                log,
+		backup:             backup,
+		namespaces:         namespaces,
+		resources:          resources,
+		labelSelector:      labelSelector,
+		dynamicFactory:     dynamicFactory,
+		discoveryHelper:    discoveryHelper,
+		backedUpItems:      backedUpItems,
+		actions:            actions,
+		podCommandExecutor: podCommandExecutor,
+		tarWriter:          tarWriter,
+		resourceHooks:      resourceHooks,
+
+		cohabitatingResources: map[string]*cohabitatingResource{
+			"deployments":     newCohabitatingResource("deployments", "extensions", "apps"),
+			"networkpolicies": newCohabitatingResource("networkpolicies", "extensions", "networking.k8s.io"),
+		},
+
+		itemBackupperFactory: &defaultItemBackupperFactory{},
+	}
+}
+
+type resourceBackupper interface {
+	backupResource(group *metav1.APIResourceList, resource metav1.APIResource) error
+}
+
+type defaultResourceBackupper struct {
+	log                *logrus.Entry
+	backup             *api.Backup
+	namespaces         *collections.IncludesExcludes
+	resources          *collections.IncludesExcludes
+	labelSelector      string
+	dynamicFactory     client.DynamicFactory
+	discoveryHelper    discovery.Helper
+	backedUpItems      map[itemKey]struct{}
+	actions            map[schema.GroupResource]Action
+	podCommandExecutor podCommandExecutor
+	tarWriter          tarWriter
+	resourceHooks      []resourceHook
+
+	cohabitatingResources map[string]*cohabitatingResource
+
+	itemBackupperFactory itemBackupperFactory
+}
+
+// backupResource backs up all the objects for a given group-version-resource.
+func (rb *defaultResourceBackupper) backupResource(
+	group *metav1.APIResourceList,
+	resource metav1.APIResource,
+) error {
+	var errs []error
+
+	gv, err := schema.ParseGroupVersion(group.GroupVersion)
+	if err != nil {
+		return errors.Wrapf(err, "error parsing GroupVersion %s", group.GroupVersion)
+	}
+	gr := schema.GroupResource{Group: gv.Group, Resource: resource.Name}
+	grString := gr.String()
+
+	log := rb.log.WithField("groupResource", grString)
+
+	switch {
+	case rb.backup.Spec.IncludeClusterResources == nil:
+		// when IncludeClusterResources == nil (auto), only directly
+		// back up cluster-scoped resources if we're doing a full-cluster
+		// (all namespaces) backup. Note that in the case of a subset of
+		// namespaces being backed up, some related cluster-scoped resources
+		// may still be backed up if triggered by a custom action (e.g. PVC->PV).
+		if !resource.Namespaced && !rb.namespaces.IncludeEverything() {
+			log.Infof("Skipping resource because it's cluster-scoped and only specific namespaces are included in the backup", grString)
```

Comment:
missing format string verb

Reply:
Oops, need to remove the arg (it's in the structured log fields)